be/btree.c

```diff
@@ -262,39 +262,82 @@ static bool btree_backlink_invariant(const struct m0_be_btree_backlink *h,
 static bool btree_node_invariant(const struct m0_be_btree *btree,
 				 const struct m0_be_bnode *node, bool root)
 {
-	return
-		_0C(node->bt_header.hd_magic != 0) &&
-		_0C(m0_format_footer_verify(&node->bt_header, true) == 0) &&
-		_0C(btree_backlink_invariant(&node->bt_backlink, btree->bb_seg,
-					     &btree->bb_cookie_gen)) &&
-		_0C(node->bt_level <= BTREE_HEIGHT_MAX) &&
-		_0C(memcmp(&node->bt_backlink, &btree->bb_backlink,
-			   sizeof node->bt_backlink) == 0) &&
-		/* Expected occupancy. */
-		_0C(ergo(root, 0 <= node->bt_num_active_key &&
-			 node->bt_num_active_key <= KV_NR)) &&
-		_0C(ergo(!root, BTREE_FAN_OUT - 1 <= node->bt_num_active_key &&
-			 node->bt_num_active_key <= KV_NR)) &&
-		_0C(m0_forall(i, node->bt_num_active_key,
-			      node->bt_kv_arr[i].btree_key != NULL &&
-			      node->bt_kv_arr[i].btree_val != NULL &&
-			      m0_be_seg_contains(btree->bb_seg,
-						 node->bt_kv_arr[i].
-						 btree_key) &&
-			      m0_be_seg_contains(btree->bb_seg,
-						 node->bt_kv_arr[i].
-						 btree_val))) &&
-		_0C(ergo(!node->bt_isleaf,
-			 m0_forall(i, node->bt_num_active_key + 1,
-				   node->bt_child_arr[i] != NULL &&
-				   m0_be_seg_contains(btree->bb_seg,
-						      node->
-						      bt_child_arr[i])))) &&
-		/* Keys are in order. */
-		_0C(ergo(node->bt_num_active_key > 1,
-			 m0_forall(i, node->bt_num_active_key - 1,
-				   key_gt(btree, node->bt_kv_arr[i+1].btree_key,
-					  node->bt_kv_arr[i].btree_key))));
+	if(btree->bb_ops->ko_type == M0_BBT_CAS_CTG || btree->bb_ops->ko_type == M0_BBT_CONFDB 
+			|| btree->bb_ops->ko_type == M0_BBT_INVALID|| btree->bb_ops->ko_type == M0_BBT_UT_KV_OPS
+			|| btree->bb_ops->ko_type == M0_BBT_NR || btree->bb_ops->ko_type == M0_BBT_COB_FILEATTR_EA 
+			|| btree->bb_ops->ko_type == M0_BBT_COB_NAMESPACE || btree->bb_ops->ko_type == M0_BBT_BALLOC_GROUP_EXTENTS
+			|| btree->bb_ops->ko_type == M0_BBT_BALLOC_GROUP_DESC || btree->bb_ops->ko_type == M0_BBT_EMAP_EM_MAPPING)
```

Comment:
This piece of code is duplicated many times. Please define it as a function.

Reply:
Ok.
I will look into it.